test/unit/Cardano/WalletSpec.hs

```diff
@@ -0,0 +1,598 @@
+        ShowFmt (new block)
+    new b = flip evalState s $ do
+        let txs = transactions b
+        utxo' <- (foldMap utxoFromTx txs `restrictedTo`) <$> state (txOutsOurs txs)
+        return $ utxo' `excluding` txIns txs
+    updateUTxO' b u = evalState (updateUTxO b u) s
+
+
+prop_applyBlockBasic
+    :: WalletState
+    -> Property
+prop_applyBlockBasic s =
+    cover 90 cond0 "ours ≠ ∅ " $
+    cover 90 cond1 "addresses \\ ours ≠ ∅ " $
+        property prop
+  where
+    cond0 = not $ null $ ourAddresses s
+    cond1 = not $ null $ addresses \\ (ourAddresses s)
+    prop =
+        let
+            checkpoints = initWallet s :| []
+            (wallet :| _) = foldl (flip applyBlock) checkpoints blockchain
+            utxo = totalUTxO wallet
+            utxo' = evalState (foldM (flip updateUTxO) mempty blockchain) s
+        in
+            (ShowFmt utxo === ShowFmt utxo') .&&.
+            (availableBalance wallet === balance utxo') .&&.
+            (totalBalance wallet === balance utxo')
+
+
+{-------------------------------------------------------------------------------
+               Basic Model - See Wallet Specification, section 3
+
+   Our implementation of 'applyBlock' is a bit more complexe than the basic
+   model. In practice, we do not want to compute intersection and tx id of a
+   whole block of transactions, but we only do it for the one that are relevant
+   to us.
+   Plus, we are tracking more than just the UTxO. However, when it comes to UTxO
+   the basic model and our implementation should be "on-par" and therefore,
+   given a few blocks, we should be able to control that they are indeed.
+-------------------------------------------------------------------------------}
+
+-- Update UTxO as described in the formal specification, Fig 3. The basic model
+updateUTxO
+    :: IsOurs s
+    => Block
+    -> UTxO
+    -> State s UTxO
+updateUTxO !b utxo = do
+    let txs = transactions b
+    utxo' <- (foldMap utxoFromTx txs `restrictedTo`) <$> state (txOutsOurs txs)
+    return $ (utxo <> utxo') `excluding` txIns txs
+
+
+{-------------------------------------------------------------------------------
+                                  Test Data
+
+    In practice, we may want to generate arbitrary valid sequences of block.
+    This isn't trivial though because we would need to generate _valid_ chains
+    for various invariants and preconditions to hold. Work have been done in
+    cardano-sl to generate such chains, and we may want to use that at some
+    point. For now, a valid chain coming from the testnet will do
+
+-------------------------------------------------------------------------------}
+
+-- | An arbitrary wallet state that can recognize some hard-coded addresses from
+-- our chain. This allows us to control that the UTxO gets updated accordingly
+-- for some arbitrary instances of that state.
+data WalletState = WalletState
+    { ourAddresses :: [Address]
+    , discoveredAddresses :: Set Address
```

Comment:
why `discoveredAddresses` is Set and `ourAddresses` is List?

Reply:
To leverage QuickCheck built-in for shrinking. We don't shrink the `discoveredAddresses` because we don't generate initial ones in the generator and we may add the same one multiple times as we discover (hence the set). `ourAddresses` doesn't change however as we apply blocks. 
Seems confusing though, I could also just `toList` and `fromList` in the shrinker :+1:

Reply:
Fixed.